Resolve circular dependency

neelasha23 opened the pull request:
## Describe your changes
1. Moved the `store` utility functions like getting all saved snippets, finding dependents of key, etc to `store_utils.py`, and corresponding tests moved to `test_store.py`.
2. All util functions that access current connection moved from `util.py` to `inspect.py`, and corresponding tests moved to `test_inspect.py`
3. Renamed `error_messages.py` to `error_handler.py` and structured the code by adding more internal functions.
4. Moved logic of extracting table names from query to `util.py`.
5. Added functions to check if error is sqlalchemy error or not, and substring check in `util.py`. this is needed for the detailed error message logic.
6. Handling of DuckDB Parser error and pyodbc.ProgrammingError added. so some of the integration tests failing previously have now been fixed.

## Issue number

Closes #749  #545 

## Checklist before requesting a review

- [ ] Performed a self-review of my code
- [ ] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [ ] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [ ] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)



<!-- readthedocs-preview jupysql start -->
----
:books: Documentation preview :books:: https://jupysql--787.org.readthedocs.build/en/787/

<!-- readthedocs-preview jupysql end -->

neelasha23 commented:
Please review @edublancas

edublancas commented:
@neelasha23 I was about to review this but now I see that the tests are failing. what's the status? I guess there are merge conflicts due to recent changes in master?

neelasha23 commented:
> @neelasha23 I was about to review this but now I see that the tests are failing. what's the status? I guess there are merge conflicts due to recent changes in master?

Yes the snippets related tests started failing after I rebased. Fixed now. Please review.
@edublancas

neelasha23 commented:
Please review @edublancas

edublancas reviewed (changes requested):
I just realized that I had reviewed this before but never submitted my observations  (they still appeared as "pending" but I think that means you'll never saw them), so I'm submitting them now.

there's a lot of changes in this PR so please write a brief description so I know what to expect. e.g., "moved functions from a.py, b.py, moved functions from c.py into a separate module"

i guess this is mostly moving stuff around and not adding any new stuff so please explain

neelasha23 commented on src/sql/store_utils.py:
```diff
@@ -0,0 +1,49 @@
+from sql.store import store, get_dependents_for_key
```
yes when we add these functions in `store.py` the tests fail. In `isolate_tests` fixture there's the store is being cleared: `store.store = store.SQLStore()`. This is causing the store to become empty when running each test, although we are creating snippets again in the test.
I'm not sure exactly why this is happening but moving to a separate file solves the issue.

neelasha23 commented on src/tests/test_inspect.py:
```diff
@@ -89,6 +96,48 @@ def test_get_column(sample_db, name, first, second, schema):
     assert second in columns._repr_html_()
 
 
+@pytest.mark.parametrize(
```
yes these are old tests moved from `test_util.py`

neelasha23 commented on src/tests/test_magic_cte.py:
```diff
@@ -63,7 +63,11 @@ def test_infer_dependencies(ip, capsys):
 Did you mean : 'author_sub'
-If you need help solving this issue, send us a message: https://ploomber.io/community
+
+
+Original error message from DB driver:
+(sqlite3.OperationalError) no such table: author_subb
+[SQL: SELECT last_name FROM author_subb;]
 """
```
previously it was in `magic.py` due to the circular dependency. Now it's move to the error handler file and similar to the syntax error it displays descriptive error message + original db error. original db error was not there previously.

neelasha23 commented on CHANGELOG.md:
```diff
@@ -12,6 +12,8 @@
 * [Fix] `ResultSet` footer only displayed when `feedback=2`
 * [Fix] Current connection and switching connections message only displayed when `feedback>=1`
 * [Fix] `--persist/--persist-replace` perform `ROLLBACK` automatically when needed
+* [Fix] Resolve circular dependency issue occurring due to multiple imports in `util.py`
```
[fixed](https://github.com/ploomber/jupysql/pull/787/commits/10a987fd245032cae516c5719ec4ebf5c5f77d32)

neelasha23 commented on CHANGELOG.md:
```diff
@@ -12,6 +12,8 @@
 * [Fix] Current connection and switching connections message only displayed when `feedback>=1`
 * [Fix] `--persist/--persist-replace` perform `ROLLBACK` automatically when needed
+* [Fix] Resolve circular dependency issue occurring due to multiple imports in `util.py`
+* [Fix] Handling of DuckDB `Parser Error` and `pyodbc.ProgrammingError`
```
[fixed](https://github.com/ploomber/jupysql/pull/787/commits/10a987fd245032cae516c5719ec4ebf5c5f77d32)

neelasha23 commented on src/tests/test_inspect.py:
```diff
@@ -11,6 +11,13 @@
 from sql import inspect, connection
 
 
+EXPECTED_SUGGESTIONS_MESSAGE = "Did you mean :"
```
[fixed](https://github.com/ploomber/jupysql/pull/787/commits/10a987fd245032cae516c5719ec4ebf5c5f77d32)

neelasha23 commented on src/sql/connection/connection.py:
```diff
@@ -858,11 +858,7 @@ def _start_sqlalchemy_connection(cls, engine, connect_str):
-                raise exceptions.UsageError(detailed_msg)
-            else:
-                print(e)
+            ErrorHandler(e).handle_exception()
```
[fixed](https://github.com/ploomber/jupysql/pull/787/commits/587279506d14c0d97a70f41c0efab07f7029eb7c)

neelasha23 commented:
Addressed your comments and updated the PR description @edublancas

edublancas commented:
@neelasha23 good work!

when writing your initial comment, write each issue on a separate line:

closes #X
closes #Y

because if you do

closes #X #Y

github won't close both issues
